components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Services/SubscriptionServices.cs

```diff
@@ -1287,6 +1330,67 @@ private class MonitoredItemWrapper {
                 return eventFilter;
             }
 
+            public void ProcessMonitoredItemNotification(SubscriptionNotificationModel message, MonitoredItemNotificationModel monitoredItemNotification) {
+                var pendingAlarmsOptions = EventTemplate?.PendingAlarms;
+                if (pendingAlarmsOptions?.IsEnabled == true && monitoredItemNotification.Value.GetValue(typeof(EncodeableDictionary)) is EncodeableDictionary values) {
+                    var conditionIdIndex = pendingAlarmsOptions.ConditionIdIndex ?? -1;
+                    var retainIndex = pendingAlarmsOptions.RetainIndex ?? -1;
+                    if (conditionIdIndex != -1) {
+                        var conditionId = values[conditionIdIndex].Value.ToString();
+                        if (!PendingAlarmEvents.ContainsKey(conditionId) || retainIndex == -1) {
+                            PendingAlarmEvents[conditionId] = monitoredItemNotification;
+                            pendingAlarmsOptions.Dirty = true;
+                        }
+                        else {
+                            var existingValues = PendingAlarmEvents[conditionId].Value.GetValue(typeof(EncodeableDictionary)) as EncodeableDictionary;
+                            if (existingValues[retainIndex] != values[retainIndex]) {
+                                PendingAlarmEvents[conditionId] = monitoredItemNotification;
```

Comment:
We probably should set it to the last notification regardless. Even though this will not qualify as Dirty.

Reply:
Yes, very true - missed that.